File: common/src/main/java/net/onelitefeather/pandorascluster/api/PandorasClusterApiImpl.java
package net.onelitefeather.pandorascluster.api;

import net.onelitefeather.pandorascluster.api.service.*;
import net.onelitefeather.pandorascluster.database.service.*;
import org.hibernate.HibernateException;
import org.hibernate.cfg.Configuration;

public class PandorasClusterApiImpl implements PandorasCluster, ThreadHelper {

    private DatabaseService databaseService;
    private LandPlayerService landPlayerService;
    private LandFlagService landFlagService;
    private LandService landService;
    private StaffNotificationService staffNotificationService;

    public PandorasClusterApiImpl() {

        syncThreadForServiceLoader(() -> {
            try {
                var sessionFactory = new Configuration().configure().configure("connection.cfg.xml").buildSessionFactory();
                this.databaseService = new DatabaseServiceImpl(sessionFactory);
            } catch (HibernateException e) {
                this.databaseService = null;
                throw new HibernateException("Cannot build session factorty.", e);
            }
        });

        if (databaseService == null) return;
        this.landPlayerService = new DatabaseLandPlayerService(databaseService);
        this.landFlagService = new DatabaseLandFlagService(this);
        this.landService = new DatabaseLandService(this);
        this.staffNotificationService = new StaffNotificationService();
    }

    @Override
    public LandPlayerService getLandPlayerService() {
        return this.landPlayerService;
    }

    @Override
    public DatabaseService getDatabaseService() {
        return this.databaseService;
    }

    @Override
    public LandService getLandService() {
        return this.landService;
    }

    @Override
    public LandFlagService getLandFlagService() {
        return this.landFlagService;
    }

    @Override
    public StaffNotificationService getStaffNotification() {
        return this.staffNotificationService;
    }
}

File: common/src/main/kotlin/net/onelitefeather/pandorascluster/api/PandorasClusterApiImpl.kt
package net.onelitefeather.pandorascluster.api

import net.onelitefeather.pandorascluster.api.service.*
import net.onelitefeather.pandorascluster.database.service.DatabaseLandFlagService
import net.onelitefeather.pandorascluster.database.service.DatabaseLandPlayerService
import net.onelitefeather.pandorascluster.database.service.DatabaseLandService
import net.onelitefeather.pandorascluster.database.service.DatabaseServiceImpl

class PandorasClusterApiImpl : PandorasClusterApi {

    private var databaseService: DatabaseService = DatabaseServiceImpl()
    private lateinit var landPlayerService: LandPlayerService
    private lateinit var landFlagService: LandFlagService
    private lateinit var landService: LandService
    private lateinit var staffNotification: StaffNotificationService

    init {
        databaseService.connect("connection.cfg.xml")
        if (databaseService.isRunning()) {
            landService = DatabaseLandService(this, databaseService)
            landFlagService = DatabaseLandFlagService(databaseService, landService)
            landPlayerService = DatabaseLandPlayerService(databaseService, landService)
            staffNotification = StaffNotificationService(this)
        }
    }

    override fun getDatabaseStorageService(): LandService = landService

    override fun getLandPlayerService(): LandPlayerService = landPlayerService

    override fun getDatabaseService(): DatabaseService = databaseService

    override fun getLandService(): LandService = landService

    override fun getLandFlagService(): LandFlagService = landFlagService

    override fun getStaffNotification(): StaffNotificationService = staffNotification


}